Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+      _ ≤ 1 := (μ_limsup_le_one μ hs_le hψ_lim)
+      _ = smoothingFun μ x ^ a := by rw [ha, rpow_zero]
+  · have ha_pos : 0 < a := lt_of_le_of_ne a_in.1 (Ne.symm ha)
+    have h_eq : (fun n : ℕ =>
+        (μ (x ^ s (ψ n)) ^ (1 / (s (ψ n) : ℝ))) ^ ((s (ψ n) : ℝ) / (ψ n : ℝ))) =ᶠ[atTop]
+        fun n : ℕ => μ (x ^ s (ψ n)) ^ (1 / (ψ n : ℝ)) := by
+      have h : (fun n : ℕ => (1 : ℝ) / (s (ψ n) : ℝ) * (s (ψ n) : ℝ)) =ᶠ[atTop] 1 := by
+        apply Filter.EventuallyEq.div_mul_cancel_atTop
+        exact Tendsto.num (tendsto_natCast_atTop_atTop.comp hψ_mono.tendsto_atTop) ha_pos hψ_lim
+      simp_rw [← rpow_mul (apply_nonneg μ _), mul_div]
+      exact EventuallyEq.comp₂ EventuallyEq.rfl HPow.hPow (h.div EventuallyEq.rfl)
+    exact le_of_eq (Tendsto.limsup_eq (Tendsto.congr' h_eq
+      ((((tendsto_smoothingFun_of_map_one_le_one μ hμ1 x).comp
+      ((tendsto_natCast_atTop_iff (R := ℝ)).mp <|
+        Tendsto.num (tendsto_natCast_atTop_atTop.comp hψ_mono.tendsto_atTop)
+          ha_pos hψ_lim)).rpow hψ_lim (Or.inr ha_pos)))))
+
+/-- If `μ 1 ≤ 1` and `μ` is nonarchimedean, then `smoothingFun μ` is nonarchimedean. -/
+theorem isNonarchimedean_smoothingFun (hμ1 : μ 1 ≤ 1) (hna : IsNonarchimedean μ) :
+    IsNonarchimedean (smoothingFun μ) := by
+  intro x y
+  have hn : ∀ n : ℕ, ∃ m < n + 1,
+        μ ((x + y) ^ (n : ℕ)) ^ (1 / (n : ℝ)) ≤ (μ (x ^ m) * μ (y ^ (n - m : ℕ))) ^ (1 / (n : ℝ)) :=
+    fun n => RingSeminorm.exists_index_pow_le μ hna x y n
+  set mu : ℕ → ℕ := fun n => mu μ hn n
+  set nu : ℕ → ℕ := fun n => n - mu n with hnu
+  have hmu_le : ∀ n : ℕ, mu n ≤ n := fun n => mu_le μ hn n
+  have hmu_bdd : ∀ n : ℕ, (mu n : ℝ) / n ∈ Set.Icc (0 : ℝ) 1 := fun n => mu_bdd μ hn n
+  have hs : Bornology.IsBounded (Set.Icc (0 : ℝ) 1) := Metric.isBounded_Icc 0 1
+  obtain ⟨a, a_in, ψ, hψ_mono, hψ_lim⟩ := tendsto_subseq_of_bounded hs hmu_bdd
+  rw [closure_Icc] at a_in
+  set b := 1 - a with hb
+  have hb_lim : Tendsto ((fun n : ℕ => (nu n : ℝ) / ↑n) ∘ ψ) atTop (𝓝 b) := by
+    apply Tendsto.congr' _ (Tendsto.const_sub 1 hψ_lim)
+    simp only [EventuallyEq, Function.comp_apply, eventually_atTop, ge_iff_le]
+    use 1
+    intro m hm
+    have h0 : (ψ m : ℝ) ≠ 0 := cast_ne_zero.mpr (_root_.ne_of_gt (lt_of_le_of_lt (_root_.zero_le _)
+      (hψ_mono (Nat.pos_of_ne_zero (one_le_iff_ne_zero.mp hm)))))
+    rw [← div_self h0, ← sub_div, cast_sub (hmu_le _)]
+  have b_in : b ∈ Set.Icc (0 : ℝ) 1 := unitInterval.mem_iff_one_sub_mem.mp a_in
+  have hnu_le : ∀ n : ℕ, nu n ≤ n := fun n => by simp only [hnu, tsub_le_self]
+  have hx : limsup (fun n : ℕ => μ (x ^ mu (ψ n)) ^ (1 / (ψ n : ℝ))) atTop ≤
+      smoothingFun μ x ^ a := limsup_mu_le μ hμ1 hmu_le a_in hψ_mono hψ_lim
+  have hy : limsup (fun n : ℕ => μ (y ^ nu (ψ n)) ^ (1 / (ψ n : ℝ))) atTop ≤
+      smoothingFun μ y ^ b :=
+    limsup_mu_le μ hμ1 hnu_le b_in hψ_mono hb_lim
+  have hxy : limsup
+      (fun n : ℕ => μ (x ^ mu (ψ n)) ^ (1 / (ψ n : ℝ)) * μ (y ^ nu (ψ n)) ^ (1 / (ψ n : ℝ))) atTop ≤
```

Comment:
This line seems to be longer than 100 chars, no?

Reply:
It is exactly 100 chars.

Reply:
I can shorten it a bit by removing the type of `n`.

Reply:
It should be my VSCode then (I hope you had not actually _counted_ them). Yes, removing the type seems a good idea.

Reply:
I used regular expression search in VSCode to search for `.{100}`. In any case, I think a linter would complain if it were over 100 chars (but I removed the type anyways).